core/matrix/ell.hpp

```diff
@@ -0,0 +1,268 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright 2017-2018
+
+Karlsruhe Institute of Technology
+Universitat Jaume I
+University of Tennessee
+
+Redistribution and use in source and binary forms, with or without modification,
+are permitted provided that the following conditions are met:
+
+1. Redistributions of source code must retain the above copyright notice,
+   this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright notice,
+   this list of conditions and the following disclaimer in the documentation
+   and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its contributors
+   may be used to endorse or promote products derived from this software
+   without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_CORE_MATRIX_ELL_HPP_
+#define GKO_CORE_MATRIX_ELL_HPP_
+
+
+#include "core/base/array.hpp"
+#include "core/base/convertible.hpp"
+#include "core/base/lin_op.hpp"
+#include "core/base/mtx_reader.hpp"
+
+
+namespace gko {
+namespace matrix {
+
+template <typename ValueType>
+class Dense;
+
+/**
+ * Ell is a matrix format which aligned all nonzero elements to the left
+ * of the matrix (from the first column) and save the smaller matrix up
+ * to the maximum number of nonzero elements in a single element.
+ *
+ * The nonzero elements are stored in a 1D array column-wise, and accompanied
+ * with a column index array index of each row.
+ * An additional value max_nnz_row is also stored to identify the size.
+ * (Column) padding is also introduced to improve the performance.
```

Comment:
This is my attempt of a description:

```c++
/**
 * ELL is a compressed format which stores only the nonzero elements of the
 * matrix. All zeros are removed from each row and the resulting arrays of
 * values are stored in a _number of rows_ - by - _maximum number of nonzeros
 * per row_ table. The original column indexes of the values are stored in an
 * accompanying column index table. Both tables are stored in memory in
 * column-major order, and appropriate padding is introduced to ensure optimal
 * memory access.
 *
 * Unused elements in both tables are initialized to 0.
```
@ginkgo-project/developers @yanjen @yhmtsai  feel free to modify and improve it.

Reply:
Fixed